### ISSUE TITLE
pyffi is a 20k-line shadow library: move fit/OOS orchestration into gam behind typed entries; bindings marshal only

### ISSUE THREAD

**Issue by SauersML** [SauersML/gam#2236]
# pyffi is a 20k-line shadow library: move fit orchestration into gam, make every binding a marshaller

## The problem (measured)

Three pyffi files total ~19,859 lines of Rust: latent_basis_and_sae_ffi.rs (7,262), latent_basis_and_sae_ffi_tail.rs (3,500), geometry_ffi.rs (9,097). This is not marshalling — it is ORCHESTRATION living in the binding:

- `sae_manifold_fit_inner` (~1,200 lines): builds atoms/seeds, constructs SaeManifoldOuterObjective, runs OuterProblem, applies the top-k projection split (#1231/#1232), runs the structured-residual outer alternation (#2021) INCLUDING its Λ nursery/promotion logic, assembles the payload dict.
- `predict_oos_from_arrays` (~600 lines): rebuilds atom plans, seeds coords + logits, assembles ρ*, runs the fixed-decoder solve, applies the top-k cap.
- Plus the hybrid-image attach, encode-atlas glue, steer bundle rebuilds.

Consequences: the CLI cannot fit what python fits; a Rust library user cannot fit what python fits; every behavior fix must be tested through a wheel; the single-source-of-truth principle is violated at the largest possible scale (the fit itself). Precedent for the failure mode: verdict-string mapping was born in pyffi today and had to be moved onto the enum (ba57254af); the OOS cold-arm discriminator bug and the #871 dead-code guard are both artifacts of logic straddling the boundary.

## The rule

gam-pyffi may contain ONLY: PyReadonly→view marshalling, py.detach wrapping, error mapping, and payload↔dict conversion. Every decision (seeding policy, projection semantics, alternation schedules, verdict names, defaults) lives in gam/gam-sae behind a typed entry the CLI, Rust users, and python all call.

## Target shape

```rust
// gam-sae (or gam::terms::sae): the ONE fit entry
pub struct SaeFitRequest { /* typed config: shapes, topology, assignment, schedules, budgets */ }
pub struct SaeFitReport  { /* fitted payload: atoms, assignments, loss, termination, diagnostics — serde */ }
pub fn run_sae_manifold_fit(x: ArrayView2<f64>, req: &SaeFitRequest) -> Result<SaeFitReport, String>;
pub fn run_sae_manifold_predict_oos(x: ArrayView2<f64>, model: &SaeFitReport /* or payload */, opts: &OosOptions) -> Result<OosReport, String>;
```

`SaeFitReport` should BE (or embed) the #2091 ManifoldSaePayload — the serde payload the pyclass port already defined is exactly the single-source carrier; python's ManifoldSAE and the CLI both read it.

## Increments (each green)

1. **Pure move**: extract `sae_manifold_fit_inner`'s post-validation body into `gam-sae::manifold::fit_entry::run_sae_manifold_fit_from_parts(...)` with the same (long) argument list; pyffi calls it. No behavior change; the binding shrinks by ~1,200 lines. Same for `predict_oos_from_arrays`.
2. **Typed request/report**: replace the 40-arg signatures with SaeFitRequest/SaeFitReport (+ OosOptions/OosReport); report embeds ManifoldSaePayload + termination (#2235) + diagnostics. pyffi becomes kwargs→request and report→dict.
3. **Re-point the sibling lanes**: sae_manifold_fit_minimal / _stagewise / _ibp / the ManifoldSaeCore methods all call the library entry (delete their duplicated orchestration).
4. **CLI**: examples/dictionary CLI (or a small gam-cli) drives run_sae_manifold_fit directly — the proof the library is self-sufficient.
5. Guard: a lint-style test greps gam-pyffi for banned patterns (OuterProblem::new, seed_coords_by_decoder_projection, ...) so orchestration cannot creep back into the binding.

## Relation
- #2091 (pyclass + serde payload): the report type. #2232 (single entry): this is its FFI face. #2011 (thin-wrapper sweep): same doctrine, python side. The #871 dead-code guard becomes unnecessary for items the library entry owns (nothing is "only used by pyffi" anymore).

**Comment by SauersML** [comment]
Increment landed (403481a9b): the cold-start seed-routing policy cluster — sae_decoder_lsq_init, sae_residual_seed_logits, sae_output_energy_cluster_labels, sae_refine_periodic_seed_coords_by_cluster, sae_em_refine_routing_seed (#629/#630), and the shared SAE_OOS_PROJECTION_GRID_RESOLUTION — moved from gam-pyffi into gam-sae::manifold::seed_routing as pub typed entries (byte-identical bodies; binding reaches them via the ffi_prelude re-export). oos_entry's private duplicate of the grid-resolution const is deduped onto the shared one, and the #174/#629 regression tests moved next to the code. ~730 lines out of the binding.

Next increments: the evaluator-rebuild + width-policy cluster (build_sae_basis_evaluators, torus/cylinder/euclidean width recovery, Duchon m policy), the fisher row-metric builder, then the atom-plan build (sae_build_atom_plans / sae_manifold_build_inputs) and the payload atom_plans policy.

**Comment by SauersML** [comment]
Increment 2 landed (in eab98040e — the lane's edits were checkpoint-committed together with other live workspace edits): the atom width-policy + evaluator-rebuild cluster moved verbatim from gam-pyffi into gam-sae::manifold::evaluator_rebuild — build_sae_basis_evaluators, sae_duchon_atom_m, sae_periodic_basis_size / SAE_MAX_PERIODIC_HARMONICS, sae_torus_axis_basis_size, sae_euclidean_degree_for_basis_size (+ SEED/RECOVERY degree consts), sae_cylinder_harmonics_degree / SAE_CYLINDER_LINE_DEGREE, the Möbius production consts, SAE_SPHERE_BASIS_SIZE, SAE_DEFAULT_TORUS_HARMONICS. Diff vs pre-move verified: only pub visibility and one neutralized intra-doc link. ~360 more lines out of the binding.

Note for a later dedupe pass: gam-sae oos_entry.rs still carries its own private near-duplicates of this width policy (periodic_basis_size rejects H=0, different error strings; cylinder/mobius recovery keyed on persisted harmonics instead of the production convention). Unifying those onto evaluator_rebuild is a semantic reconciliation, not pure motion, so it is deliberately not part of this migration lane's commits.

Next: the seed atom-build library (SaeAtomBuildPlan, sae_build_atom_plans, per-kind sae_build_*_atom, sae_manifold_build_inputs padded-stack assembly, sae_pick_duchon_center_indices) and the fisher row-metric builder.

**Comment by SauersML** [comment]
## Core-entry extraction update

Several more binding-owned orchestration clusters have moved behind typed `gam-sae` entries:

- seed atom planning + padded basis/jacobian/penalty construction moved to `manifold::atom_build` (~740 lines removed from the two PyFFI fragments)
- Fisher row-metric validation/packing moved to `manifold::fisher_metric`
- full fit seed/base-term construction (assignment policy, evaluator/term construction, routing refinement, row metric/weights, rho seeding/shared ARD, compatibility checks) moved to `manifold::fit_seed`
- automatic minimal-fit seeding moved to `manifold::minimal_seed`
- stagewise seed construction moved to `manifold::stagewise_seed`
- fit-result `active_mask`/R² duplicate computation was deleted from PyFFI; the binding now marshals the `SaeFitReport` values

Focused source guards prohibit the moved constructors/policies from returning to PyFFI. The bulk landed across `adb9ab911`, `9ad1a2d51`, `c7c364c3e`, `a25536989`, and the scoped checkpoint `32ed0486a`. MSI passed `cargo check -p gam-sae --all-targets` on the intermediate checkpoint, including the previously stale example. Sibling-entry and CLI/direct-core closure is continuing.

**Comment by SauersML** [comment]
### MSI verification finding

Remote `cargo check -p gam-sae --all-targets` against `ee08e3990` exposed three migration fallout errors before PyFFI could compile: the top-k admission helper returned `SaeFitAdmission` where the typed seed validator promises `()`, and the new stagewise seed request over-constrained local center/registry borrows to the Fisher request lifetime (two E0597s). These are source issues, not runner noise; fixes are in progress in the typed seed layer. MSI also caught an unrelated `unused_mut` in the concurrently added crosscoder entry, now removed. I will post the next exact SHA and gate result after the fixes checkpoint.

**Comment by SauersML** [comment]
### Typed-entry closure + MSI repair

The sibling paths are now core-owned too: minimal fit delegates `build_sae_minimal_seed` → typed fit seed → `run_sae_manifold_fit`; stagewise delegates `build_sae_stagewise_seed` → core `fit_stagewise`; IBP is only a fixed-`ibp_map` forwarding wrapper; OOS marshals `SaeOosRequest` → `run_sae_manifold_oos`. A direct Rust proof example, `crates/gam-sae/examples/sae_fit.rs`, now constructs the typed seed and calls `run_sae_manifold_fit` without PyFFI (`fd5130b2d`).

The MSI findings from the prior comment are fixed: `SaeFitSeedRequest<'data, 'context>` decouples data/Fisher lifetimes from local centers/registry, and both assignment-aware admission arms normalize their typed admission receipts to `Result<(), String>` (`ef47ed7bf`). Current source/fmt checks are clean; the next MSI run will validate the consolidated HEAD.

**Comment by SauersML** [comment]
Independent closure audit found one remaining ownership seam: PyFFI was still computing the SAE promotion-alignment threshold via `beta_quantile` and injecting a callback into `SaeFitRequest`, which also forced the direct Rust example to supply a dummy callback. I am moving that threshold derivation into `gam-sae` (with the distribution primitive in the Rust math layer), removing the callback/helper, and extending the source guards before closure. This is the last live marshalling-vs-orchestration violation found by the audit; final MSI verification will cover the integrated result.

**Comment by SauersML** [comment]
### Closure-audit fix: residual-promotion policy is now core-owned

The final audit found one remaining non-marshalling seam: `gam-pyffi` still computed `sae_promotion_align_min` with a Beta quantile and injected it through `SaeFitRequest::align_min_from_rank`. That meant the supposedly self-sufficient Rust entry still needed a binding-owned numerical-policy callback; the direct Rust example carried a dummy callback when promotion was disabled.

Fixed on `main` across `18ea4fe36` plus the no-fallback follow-up `11e378cb6`:

- `beta_quantile` moved to the lower `gam-math` crate (and its focused tests moved with it)
- `gam-sae::manifold::fit_entry` now derives the #2071 Beta-null threshold internally
- `SaeFitRequest::align_min_from_rank`, the PyFFI helper, and the example's dummy callback were deleted
- the PyFFI no-reintroduction guard now rejects `sae_promotion_align_min`, `beta_quantile`, and `align_min_from_rank`
- valid rank > 1 computes the threshold directly; there is no compatibility/default fallback

Closure source checks confirm the cited PyFFI fragments contain none of `predict_oos_from_arrays`, `OuterProblem::new`, `seed_coords_by_decoder_projection`, or the removed promotion-policy symbols. `rustfmt --check` and `git diff --check` passed for the scoped files. Consolidated MSI compile/test verification is the remaining gate before closing.

**Comment by SauersML** [comment]
## #2236 closure — fit/OOS orchestration is core-owned; the binding marshals only

Re-audited against HEAD. The "shadow library" is gone: the SAE binding fragments now build a typed request and hand it to a `gam-sae` entry, and every orchestration/construction cluster the issue named lives in `crates/gam-sae/src/manifold/` (`fit_seed.rs`, `fit_entry.rs`, `fit_drivers.rs`, `atom_build.rs`, `evaluator_rebuild.rs`, `fisher_metric.rs`, `seed_routing.rs`, `minimal_seed.rs`, `stagewise_seed.rs`, …).

**Marshalling seam (binding → core):**
- fit: `sae_manifold_fit_inner` builds `SaeFitSeedRequest` → `build_sae_fit_seed(...)` → `gam::terms::sae::manifold::run_sae_manifold_fit(request)` (`crates/gam-pyffi/src/latent/latent_basis_and_sae_ffi.rs:2424`)
- OOS: `SaeOosRequest` → `gam::terms::sae::manifold::run_sae_manifold_oos(request)` (`latent_basis_and_sae_ffi_tail.rs:457`, `manifold_and_posterior_ffi.rs:6876`)

`latent_basis_and_sae_ffi.rs` dropped from the cited 7262 lines to 4207; the residue is request assembly + report marshalling, not math.

**Guards (all tagged `_2236`, in `crates/gam-pyffi/tests/src_modules/lib_tests.rs`) — verified holding on HEAD:**
- `sae_atom_construction_stays_in_gam_sae_2236` — no `sae_build_*_atom` / `SaeAtomBuildPlan` in the binding (0 hits)
- `sae_fisher_metric_construction_stays_in_gam_sae_2236` — no `row_metric_from_fisher_provenance` / `RowMetric::output_fisher` packing in the binding (0 hits)
- `sae_fit_seed_construction_stays_in_gam_sae_2236` — asserts `build_sae_fit_seed(SaeFitSeedRequest` IS present (1 hit) and seed/config/rho policy + promotion policy (`beta_quantile`/`align_min_from_rank`/`sae_promotion_align_min`) are absent (0 hits)
- `sae_fit_report_fields_are_marshaled_without_recomputation_2236` — no `rss`/`tss`/`active_mask` recompute in `fit_inner` (0 hits)
- `sae_sibling_fit_seeds_delegate_to_gam_sae_2236` — minimal/stagewise/IBP fits delegate to the typed core

Compile-green confirmed by the GHA warm wheel `gamfit-0.1.253` (manylinux_2_28) built from this tree. Closing: orchestration is core-owned behind typed entries, the binding is a pure marshaller, and the boundary is enforced by source guards.

**Comment by SauersML** [comment]
Closed after exact-SHA MSI verification at `1563dbe60cc6a97fd03b43753656b92deb5cfd54`.

Acceptance evidence:
- gam-pyffi #2236 guard tests: 5/5 passed;
- chart-coordinate core tests: 3/3 passed;
- typed OOS entry tests: 5/5 passed;
- direct Rust `sae_fit` example compiles;
- banned/deleted E-step orchestration surface scan passed;
- release PyFFI wheel built, installed, and passed the complete Torch contract/routing battery.

MSI job 12972041 finished with `VERIFY_DONE ... failures=0`. The binding is a marshalling/report surface over the typed gam-sae fit/OOS entries; the residual-promotion policy is core-owned.
